src/backend/distributed/executor/executor.c

```diff
@@ -1334,6 +1382,19 @@ ConnectionStateMachine(WorkerSession *session)
 
 				/* remove connection from wait event set */
 				execution->connectionSetChanged = true;
+
+				/*
+				 * Reset the transaction state machine since CloseConnection()
+				 * relies on it and even if we're not inside a distributed transaction
+				 * we set the transaction state (e.g., REMOTE_TRANS_SENT_COMMAND).
+				 */
+				if (!execution->isTransaction)
```

Comment:
This doesn't feel like the right solution. I guess we've overloaded the meaning of transactionState a bit. Perhaps the right solution would be to have a commandState. A less wrong solution would be to have a flag on the RemoteTransaction indicating whether or not we sent BEGIN.

Reply:
> This doesn't feel like the right solution. 

Yeah, I was expecting to get feedback to this change :)  But, are you OK with `connection->remoteTransaction.transactionState = REMOTE_TRANS_INVALID;` after we *somehow* decide to set the transaction state?

> Perhaps the right solution would be to have a commandState.

Do you mean not having `REMOTE_TRANS_SENT_BEGIN, REMOTE_TRANS_SENT_COMMAND, REMOTE_TRANS_FETCHING_RESULTS` and `REMOTE_TRANS_CLEARING_RESULTS` and instead have a `CommandState` in `RemoteTransaction`? Or, keep those states and have `commandState` in `WorkerSession`? The former seems a big change to me.

Reply:
> Do you mean not having REMOTE_TRANS_SENT_BEGIN, REMOTE_TRANS_SENT_COMMAND, REMOTE_TRANS_FETCHING_RESULTS and REMOTE_TRANS_CLEARING_RESULTS and instead have a CommandState in RemoteTransaction? 

I assume you pointed me to above.

I can see that `CommandState ` approach would make sense. But, that'd make the changes in this PR pretty big. So, for now, I'm adding the `beginSent` flag, which I'll revisit in a separate on top of this to implement `commandState`